### PR TITLE
[vm] Reduce unneeded copy for obtaining resource bytes

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation

Use `Cow` to reduce copies done in MoveVM.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

`cargo test`. Will provide more benchmark numbers shortly.

### CODE REVIEW

**Comment by sblackshear** [comment]
Do we have benchmarks to validate the performance improvement from this change?

**Comment by bmwill** [comment]
Seems these didn't get posted but I got this from the canary GHA job

Cluster Test Result
```
Test runner setup time spent 247 secs
Compatibility test results for land_3a654363 ==> land_f56a57ad (PR)
1. All instances running land_3a654363, generating some traffic on network
2. First full node land_3a654363 ==> land_f56a57ad, to validate new full node to old validator node traffic
3. First Validator node land_3a654363 ==> land_f56a57ad, to validate storage compatibility
4. First batch validators (14) land_3a654363 ==> land_f56a57ad, to test consensus and traffic between old full nodes and new validator node
5. First batch full nodes (14) land_3a654363 ==> land_f56a57ad
6. Second batch validators (15) land_3a654363 ==> land_f56a57ad, to upgrade rest of the validators
7. Second batch of full nodes (15) land_3a654363 ==> land_f56a57ad, to finish the network upgrade, time spent 674 secs
8. all up : 871 TPS, 5219 ms latency, 6000 ms p99 latency, no expired txns, time spent 250 secs
```

File: language/tools/resource-viewer/src/lib.rs [hunk 1]
```diff
@@ -300,7 +301,7 @@ impl RemoteCache for NullStateView {
         &self,
         _address: &AccountAddress,
         _tag: &StructTag,
-    ) -> PartialVMResult<Option<Vec<u8>>> {
+    ) -> PartialVMResult<Option<Cow<[u8]>>> {
```
Comment (vgao1996) [inline language/tools/resource-viewer/src/lib.rs:304]: So I can see that by using `Cow` you are potentially saving a copy of the bytes, but I wonder if a simple reference would also achieve the same goal? IIRC there's nothing in the Move VM that actually wants to modify these bytes so the "on-write" part seem redundant to me.

Comment (runtian-zhou) [reply language/tools/resource-viewer/src/lib.rs:304]: The problem is from `StateView`. The `StateView` returns a owned `Vec<u8>` on storage read. In some implementation of the `RemoteCache` there isn't this caching behavior so a `Cow` is actually needed over there.

**Comment by runtian-zhou** [comment]
Running the benchmark itself I actually didn't see a significant speed gain. The main reason is that `StateViewCache` isn't actually caching the data returned from the state view. Thus there isn't enough copies being reduced. However I do suggest we should commit this PR. It is a much relaxed api from vm's standpoint and we should try to get rid of the locations that returns `Vec<u8>` as much as possible as a next step.

**Comment by sblackshear** [comment]
> Running the benchmark itself I actually didn't see a significant speed gain. The main reason is that `StateViewCache` isn't actually caching the data returned from the state view. Thus there isn't enough copies being reduced. However I do suggest we should commit this PR. It is a much relaxed api from vm's standpoint and we should try to get rid of the locations that returns `Vec<u8>` as much as possible as a next step.

Weren't these changes suggested due to some profiling or benchmarking done by George? And if so, could we include a benchmark with instructions for replicating the results?

**Comment by runtian-zhou** [comment]
> Weren't these changes suggested due to some profiling or benchmarking done by George? And if so, could we include a benchmark with instructions for replicating the results?

The problem was that this benchmark was done when he implemented the parallel execution code. So the benchmark isn't really on main at this moment. And it seems to have few impact on the single threaded execution setup

**Comment by sblackshear** [comment]
> Running the benchmark itself I actually didn't see a significant speed gain. The main reason is that StateViewCache isn't actually caching the data returned from the state view. Thus there isn't enough copies being reduced.

> The problem was that this benchmark was done when he implemented the parallel execution code. So the benchmark isn't really on main at this moment. And it seems to have few impact on the single threaded execution setup

I'm trying to understand why the above caching behavior happens in the parallel execution code, but not in the benchmarking code. Can we add a new benchmark that exercises the caching behavior in the same way that the parallel execution code does?

**Comment by runtian-zhou** [comment]
With the comments I think I will yield on this PR. We should revisit  this PR once we have confidence that this indeed improves the perf (say when parallel execution gets implemented).

**Pull request closed without merging**